File: apps/moltnet-cli/agent_key_selection.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

type selectedAgentKey struct {
	Reference SecretReference
	TeamID    string
}

func hasAgentKeyConfiguration(config *CredentialsFile) bool {
	return config != nil && (config.AgentKeyRef != nil || len(config.AgentKeyRefs) > 0)
}

func hasUsableCredentialConfiguration(config *CredentialsFile) bool {
	return hasAgentKeyConfiguration(config) || (config != nil && config.OAuth2.ClientID != "")
}

// Select exactly once. Provider failure must never try another grant.
func selectAgentKeyReference(config *CredentialsFile, selectedTeam string) (*selectedAgentKey, error) {
	if config == nil {
		return nil, nil
	}
	team := strings.TrimSpace(selectedTeam)
	if ref, ok := config.AgentKeyRefs[team]; team != "" && ok {
		return &selectedAgentKey{ref, team}, nil
	}
	if config.AgentKeyRef != nil {
		return &selectedAgentKey{Reference: *config.AgentKeyRef}, nil
	}
	if len(config.AgentKeyRefs) == 0 {
		return nil, nil
	}
	if team == "" && len(config.AgentKeyRefs) == 1 {
		for id, ref := range config.AgentKeyRefs {
			if strings.TrimSpace(id) == "" {
				return nil, fmt.Errorf("team key map contains an empty team ID")
			}
			return &selectedAgentKey{ref, id}, nil
		}
	}
	if team != "" {
		return nil, fmt.Errorf("no agent key configured for the selected team; enroll that team first")
	}
	return nil, fmt.Errorf("multiple team agent keys configured; select a team explicitly")
}

func validateSelectedAgentKey(selection *selectedAgentKey, subjectID string) error {
	if selection == nil {
		return fmt.Errorf("no agent key selected")
	}
	if !secretProviderNamePattern.MatchString(selection.Reference.Provider) {
		return fmt.Errorf("agent key provider name is invalid")
	}
	return validateSecretReferenceBinding(credentialAgentKey, selection.Reference, credentialBindingIDs{
		SubjectID: subjectID, TeamID: selection.TeamID,
	})
}

// Commands with an explicit resource team select it before the location context.
func resolveCredentialTeam(credPath string, explicit ...string) (string, error) {
	if len(explicit) > 0 && strings.TrimSpace(explicit[0]) != "" {
		return strings.TrimSpace(explicit[0]), nil
	}
	if team := strings.TrimSpace(os.Getenv("MOLTNET_TEAM_ID")); team != "" {
		return team, nil
	}
	path, err := resolveCredentialsPath(credPath)
	if err != nil {
		return "", err
	}
	cwd, err := os.Getwd()
	if err != nil {
		return "", err
	}
	binding, err := resolveContextBinding(filepath.Dir(path), cwd)
	if err != nil {
		return "", fmt.Errorf("resolve credential team: %w", err)
	}
	return binding.teamID(), nil
}

File: libs/agent-config/src/agent-key-selection.ts
import {
  type AgentKeyConfiguration,
  agentKeyKey,
  type OAuth2Config,
  type SecretReference,
} from './config.js';

/** Presence only: a configured but invalid credential must fail resolution. */
export function hasAgentKeyConfiguration(
  config: AgentKeyConfiguration,
): boolean {
  return (
    config.agent_key_ref !== undefined ||
    Object.keys(config.agent_key_refs ?? {}).length > 0
  );
}

export function hasUsableCredentialConfiguration(
  config: AgentKeyConfiguration & { oauth2?: OAuth2Config },
): boolean {
  return hasAgentKeyConfiguration(config) || Boolean(config.oauth2?.client_id);
}

export interface SelectedAgentKey {
  reference: SecretReference;
  /** Absent for the compatibility fallback, even when a team was selected. */
  teamId?: string;
}

/** Select once, before contacting any provider. Failure never tries another grant. */
export function selectAgentKeyReference(
  config: AgentKeyConfiguration,
  selectedTeam?: string,
): SelectedAgentKey | null {
  const team = selectedTeam?.trim();
  const entries = config.agent_key_refs ?? {};
  if (team && Object.hasOwn(entries, team)) {
    return { reference: entries[team], teamId: team };
  }
  if (config.agent_key_ref !== undefined) {
    return { reference: config.agent_key_ref };
  }
  const teams = Object.keys(entries);
  if (!teams.length) return null;
  if (!team && teams.length === 1) {
    if (!teams[0].trim())
      throw new Error('Team key map contains an empty team ID');
    return { reference: entries[teams[0]], teamId: teams[0] };
  }
  throw new Error(
    team
      ? 'No agent key configured for the selected team; enroll that team first'
      : 'Multiple team agent keys configured; select a team explicitly',
  );
}

/** Provider-independent binding, also used before storing a reference. */
export function assertAgentKeyReferenceBinding(
  selection: SelectedAgentKey,
  subjectId?: string,
): void {
  const { reference, teamId } = selection;
  const subject = subjectId?.trim();
  if (!subject || (teamId !== undefined && !teamId.trim())) {
    throw new Error(
      'Agent key binding requires subject_id and a nonempty team',
    );
  }
  const expected = agentKeyKey(subject, teamId);
  if (
    !reference ||
    !/^[a-z][a-z0-9-]*$/.test(reference.provider) ||
    reference.provider === 'env' ||
    (reference.key !== expected &&
      !(
        reference.provider === 'file' &&
        reference.key === expected.replaceAll('/', '.')
      ))
  ) {
    throw new Error(
      'Agent key reference is not bound to this subject and team',
    );
  }
}
